shared-libs/rules-engine/src/rules-emitter/emitter.metal.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * @module rules-processor.metal
+ * Processes declarative configuration code in an identical manner to nools (but without nools)
+ */
+
+class Contact {
+  constructor({ contact, reports, tasks}) {
+    this.contact = contact;
+    this.reports = reports;
+    this.tasks = tasks;
+  }
+};
+Contact.prototype.tasks = 'foo';
+
+const Task = class { constructor(x) { Object.assign(this, x); }};
+const Target = class { constructor(x) { Object.assign(this, x); }};
+
+let processContainer;
+const results = { tasks: [], targets: [] };
+
+module.exports = {
+  getContact: () => Contact,
+  initialize: (settings, scope) => {
+    let rules = settings.rules.replace(/define.*\} then \{ /s, '');
```

Comment:
another option here would be to update `cht-conf` to not include the nools pre-code when `disableNools: true`

Reply:
I came here to say exactly this ^

> another option here would be to update cht-conf to not include the nools pre-code when disableNools: true

I think it's worth doing this. We can add documentation that that option is only available on certain versions of cht-core. To be even safer we should add a validation check in cht-conf when pushing the config to ensure the version is right before pushing the tasks code so it's caught before it gets deployed. Would that be enough to mitigate your issues 1 and 2?

Reply:
I believe it is incompatible with the idea below to avoid the config flag and disable nools by default.

> One idea would be that we could change cht-conf to set an isDeclarative flag when compiling/uploading. Could be in app_settings and not base_settings so not a user-facing config flag. This would mean that cht-core 4.2 + cht-conf latest would together have metal emitter as default for declarative configurations only.

Which of these is more important? @garethbowen @dianabarsan

Reply:
I think `disableNools: true` sounds very scary for someone who doesn't have complete understanding of the issue. I'm almost certain that outside a very limited group of people, `nools` is a complete black box. How do we convey to an app developer that this new flag is safe for them or not? 

Can cht-conf detect whether or not the config is declarative and cht supports non-nools emitters? If yes, omit adding the nools pre-code completely. 
Can cht-core detect whether the config has the nools pre-code? If it does, use nools emitter, if it doesn't use your new emitter? 
This would mean that we never push a pre-built config to a server without verifying compatibility, so it adds some complexity to cht-conf.

Reply:
> Can cht-conf detect whether or not the config is declarative and cht supports non-nools emitters?

cht-conf can detect whether cht support non-nools emitters but only if it connects to cht which is a pain. We could add this requirement if we decide that it's worth it.

> Can cht-core detect whether the config has the nools pre-code?

Yes but not in a backwards compatible way, ie: if you use the new cht-conf to push config without pre-code to an old cht-core then you're gonna have a bad time. A legacy mode flag would resolve this at the risk of someone forgetting to use it.

I'm leaning towards running the server compatibility check. @kennsippell How much of a pain would it be if this required a server URL? I guess it would cause test harness issues? What if the test harness gave an environment variable or something to generate the nools config that it expects?

Reply:
Check this out and let me know what you think. https://github.com/medic/cht-conf/pull/536

Devs can use `--dev` mode for the harness which runs application code without any compilation. But still I don't think compiling code should require a service-level dependency. That's sucks and is slow for something many devs do hundreds of times per day.

I've proposed adding the `isDeclarative` flag during compilation and adding the nools boilerplate selectively during upload ....